### PR TITLE
Add dynamic shape resizing for binary inputs

### PR DESCRIPTION
**Pull request by michalkulakowski**

### CODE REVIEW

File: src/binaryutils.cpp [hunk 1]
```diff
@@ -110,7 +110,7 @@ bool resizeNeeded(const cv::Mat& image, const std::shared_ptr<TensorInfo>& tenso
             rows = image.rows;
         }
     }
-    if (cols != Dimension(image.cols) || rows != Dimension(image.rows)) {
+    if ((!cols.match(Dimension(image.cols).getStaticValue())) || (!rows.match(Dimension(image.rows).getStaticValue()))) {
```
Comment (dkalinowski) [inline src/binaryutils.cpp:113]: Why do we need to convert static `image.cols` to Dimension object just to get static value back?
Wouldnt it be simpler to do:
```
cols.match(image.cols)
```

File: src/test/binaryutils_test.cpp [hunk 2]
```diff
@@ -197,4 +197,142 @@ TEST_F(BinaryUtilsTest, positive_resizing) {
...
+    ov::runtime::Tensor tensor;
+
+    std::shared_ptr<TensorInfo> tensorInfo = std::make_shared<TensorInfo>("", ovms::Precision::U8, ovms::Shape{1, -1, -1, 3}, InferenceEngine::Layout::NHWC);
+
+    ASSERT_EQ(convertStringValToBlob_2(stringVal, tensor, tensorInfo, false), ovms::StatusCode::OK);
+    shape_t tensorDims = tensor.get_shape();
+    size_t expectedRowsNumber = 1;
+    EXPECT_EQ(tensorDims[1], expectedRowsNumber);
+    size_t expectedColsNumber = 1;
+    EXPECT_EQ(tensorDims[1], expectedColsNumber);
```
Comment (dkalinowski) [inline src/test/binaryutils_test.cpp:333]: Shouldnt here be `tensorDims[2]`?

**Review by mkuczyns11 — APPROVED** [review]

**Review by rasapala — APPROVED** [review]
LGTM

File: docs/binary_input.md [hunk 3]
```diff
@@ -13,7 +13,11 @@ shape: [N] with dtype: DT_STRING. Where N represents elements with binary data c
 
 ## Preparing for deployment
 Before processing in the target AI model, binary image data is encoded by OVMS to a NHWC layout in BGR color format.
-It will also be resized to the model or pipeline node resolution.
+It will also be resized to the model or pipeline node resolution. When the model shape is dynamic and image data shape
+is out of range it will be adjusted to the nearer border. For example, when model shape is: [1,3,100:200,200]:
```
Comment (dkalinowski) [inline docs/binary_input.md:17]: When using binary input, the layout of data is in NHWC order, meaning the 3 will be at last position.

File: src/test/prediction_service_test.cpp [hunk 4]
```diff
@@ -844,6 +844,30 @@ TEST_F(TestPredict, PerformInferenceWithBinaryInputChangeModelInputLayout) {
...
+ *
+ * 1. Load model with dynamic shape and input layout=nhwc, initial internal layout: nchw
+ * 2. Do the inference with single binary image tensor with witdth exceeding shape range - expect status OK and reshaped output tensor
+ */
+TEST_F(TestPredict, PerformInferenceWithBinaryInputAndShapeDynamic) {
+    using namespace ovms;
+
+    // Prepare model with changed layout to nhwc (internal layout=nchw)
+    ModelConfig config = INCREMENT_1x3x4x5_MODEL_CONFIG;
+    config.setBatchingParams("auto");
```
Comment (dkalinowski) [inline src/test/prediction_service_test.cpp:858]: Why is it auto? We do not use this feature in this test. Model batch size is 1 and request batch size i 1 as well

File: src/test/prediction_service_test.cpp [hunk 5]
```diff
@@ -844,6 +844,30 @@ TEST_F(TestPredict, PerformInferenceWithBinaryInputChangeModelInputLayout) {
...
+    ASSERT_EQ(config.parseShapeParameter("(1,1,2:5,3)"), ovms::StatusCode::OK);
```
Comment (dkalinowski) [inline src/test/prediction_service_test.cpp:859]: Please note in the commen what is binary input width and height so this is clear that resizing to nearest border happens

**Review by dkalinowski — APPROVED** [review]

**Pull request merged**